src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,326 @@
+        _;
+    }
+
+    /**
+     * @notice Defines addresses and gives `msg.sender` admin access.
+     * @param _psm PSM contract address.
+     * @param _dai DAI contract address.
+     * @param _gem GEM contract address.
+     * @param _to RwaUrn contract address.
+     */
+    constructor(
+        address _dai,
+        address _gem,
+        address _psm,
+        address _to
+    ) public {
+        require(_to != address(0), "RwaInputConduit3/invalid-to-address");
+        require(PsmAbstract(_psm).dai() == _dai, "RwaInputConduit3/wrong-dai-for-psm");
+        require(GemJoinAbstract(PsmAbstract(_psm).gemJoin()).gem() == _gem, "RwaInputConduit3/wrong-gem-for-psm");
+
+        // We assume that DAI will alway have 18 decimals
+        to18ConversionFactor = 10**_sub(18, GemAbstract(_gem).decimals());
+
+        psm = PsmAbstract(_psm);
+        dai = DaiAbstract(_dai);
+        gem = GemAbstract(_gem);
+
+        to = _to;
+
+        // Give unlimited approval to PSM gemjoin
+        GemAbstract(_gem).approve(address(psm.gemJoin()), type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    /*//////////////////////////////////
+               Authorization
+    //////////////////////////////////*/
+
+    /**
+     * @notice Grants `usr` admin access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` admin access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"to", "quitTo", "psm"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else if (what == "to") {
+            to = data;
+        } else if (what == "psm") {
+            require(PsmAbstract(data).dai() == address(dai), "RwaInputConduit3/wrong-dai-for-psm");
+            require(
+                GemJoinAbstract(PsmAbstract(data).gemJoin()).gem() == address(gem),
+                "RwaInputConduit3/wrong-gem-for-psm"
+            );
+            psm = PsmAbstract(data);
```

Comment:
We need to approve gem allowance for the new psm (and dai allowance in the output conduit).
Also revoke approvals for the old psm.

I guess maybe also add a minimal test for that scenario of switching psms so we know it works.

Reply:
Done

Reply:
Logic change looks good.

Regarding the test, sorry for not making it clearer, but I actually meant a version of `testPush()` that just files a new psm before calling `outputConduit.push()`, and also checks the new allowance.
That way you verify things work after the switch.

And then in `testFile` you don't really need to create a new psm contract, and can just file any address as the new psm.

Reply:
Well, saying that if it fails the problem is in the PSM itself is only because you assume logic of `file` is enough, but that's exactly what we want to check.

Regarding the checks of approvals in `testFile` - ok, it makes sense.

Reply:
Ideally we should have a method for onboarding PSM, then we won't have a file + push test because this method will be used for both constructor and file(PSM).

Reply:
Maybe, but I wouldn't overcomplicate it.
The push test is pretty simple, just a small change from the original one.

Reply:
Do you want me to `file` to the same PSM or do we need to create a new one?

Reply:
I think a new one is better, as it has a different address, so any past approvals won't affect it

Reply:
Done